File: src/main/java/AoC2015_04.java
import static com.github.pareronia.aoc.IntegerSequence.Range.range;

import java.util.List;
import java.util.stream.IntStream;

import com.github.pareronia.aoc.codec.MD5;
import com.github.pareronia.aoc.solution.Sample;
import com.github.pareronia.aoc.solution.Samples;
import com.github.pareronia.aoc.solution.SolutionBase;

import lombok.RequiredArgsConstructor;

public final class AoC2015_04
        extends SolutionBase<AdventCoinsMiner, Integer, Integer> {

    private AoC2015_04(final boolean debug) {
        super(debug);
    }

    public static AoC2015_04 create() {
        return new AoC2015_04(false);
    }

    public static AoC2015_04 createDebug() {
        return new AoC2015_04(true);
    }

    @Override
    protected AdventCoinsMiner parseInput(final List<String> inputs) {
        assert inputs.size() == 1;
        return new AdventCoinsMiner(inputs.get(0));
    }

    @Override
    public Integer solvePart1(final AdventCoinsMiner miner) {
        return miner.findMd5StartingWithZeroes(5);
    }

    @Override
    public Integer solvePart2(final AdventCoinsMiner miner) {
        return miner.findMd5StartingWithZeroes(6);
    }

    @Override
    @Samples({
        @Sample(method = "part1", input = TEST1, expected = "609043"),
        @Sample(method = "part1", input = TEST2, expected = "1048970"),
    })
    public void samples() {
        super.samples();
    }

    public static void main(final String[] args) throws Exception {

        AoC2015_04.create().run();
    }

    private static final String TEST1 = "abcdef";
    private static final String TEST2 = "pqrstuv";
}
    
@RequiredArgsConstructor
final class AdventCoinsMiner {
    private final String secretKey;

    private boolean checkZeroes(final byte[] digest, final int zeroes) {
        int cnt = 0;
        for (final int j : range(zeroes / 2 + zeroes % 2)) {
            final byte c = digest[j];
            if ((c & 0xF0) != 0) {
                break;
            }
            cnt++;
            if ((c & 0x0F) != 0) {
                break;
            }
            cnt++;
        }
        return cnt == zeroes;
    }
    
    public int findMd5StartingWithZeroes(final int zeroes) {
        return IntStream.iterate(1, i -> i + 1)
            .dropWhile(i -> {
                final String data = this.secretKey + String.valueOf(i);
                return !checkZeroes(MD5.md5(data), zeroes);
            })
            .findFirst().getAsInt();
    }
}

File: src/main/julia/AoC2015_04.jl
#! /usr/bin/env julia
#

include("aoc_main.jl")
if abspath(PROGRAM_FILE) == @__FILE__
    include("aocd.jl")
    using .Aocd
end

module AoC2015_04

include("aoc.jl")

using MD5

function checkZeroes(digest, zeroes)
    chars = zeroes ÷ 2 + zeroes % 2
    cnt = 0
    for j ∈ 1:chars
        ch::UInt8 = digest[j][1]
        if (ch & 0xF0) == 0
            cnt += 1
            if (ch & 0x0F) == 0
                cnt += 1
                continue
            end
        end
        break
    end
    return cnt == zeroes
end

function solve(input, zeroes)
    i = 1
    while true
        checkZeroes(md5(input * string(i)), zeroes) && return i
        i += 1
    end
end

function part1(input)
    return solve(input[1], 5)
end

function part2(input)
    return solve(input[1], 6)
end

TEST1 = @aoc_splitlines("abcdef")
TEST2 = @aoc_splitlines("pqrstuv")

function samples()
    @assert part1(TEST1) == 609_043
    @assert part1(TEST2) == 1_048_970
end

end # module AoC2015_04

aoc_main(@__FILE__, ARGS, 2015, 4)
